src/EIP712.sol

```diff
@@ -5,17 +5,24 @@ import {IERC5267} from "@openzeppelin/contracts/interfaces/IERC5267.sol";
 import {MessageHashUtils} from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
 import {IEIP712} from "./interfaces/IEIP712.sol";
 import {BaseAuthorization} from "./BaseAuthorization.sol";
+import {PrefixedSaltLib} from "./libraries/PrefixedSaltLib.sol";
 
 /// @title EIP712
 /// @dev This contract does not cache the domain separator and calculates it on the fly since it will change when delegated to.
 /// @notice It is not compatible with use by proxy contracts since the domain name and version are cached on deployment.
 contract EIP712 is IEIP712, IERC5267, BaseAuthorization {
+    using PrefixedSaltLib for uint96;
     /// @dev `keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract,bytes32 salt)")`.
+
     bytes32 internal constant _DOMAIN_TYPEHASH = 0xd87cd6ef79d4e2b95e15ce8abf732db51ec771f1ca2edccf22a46c729ac56472;
 
     /// @dev Cached name and version hashes for cheaper runtime gas costs.
     bytes32 private immutable _cachedNameHash;
     bytes32 private immutable _cachedVersionHash;
+    address private immutable _cachedImplementation;
+
+    /// @dev Any prefix to be added to the salt. This can be updated by the owner but it defaults to 0.
+    uint96 private _saltPrefix;
```

Comment:
what if this was 

uint256 private _cachedFullSalt;

and anytime you update the salt you just save it already packed with the _cachedImplementation

then you can just use _cachedFullSalt in the function call to eip712Domain() and call to domainSeparator()

then you only have to do .pack() once

Reply:
Unfortuantely we can't set storage in constructor for delegated contracts so I think this is the best approach still